src/VisualStudio/Core/Def/RoslynPackage.cs

```diff
@@ -151,60 +152,49 @@ protected override async Task InitializeAsync(CancellationToken cancellationToke
 
-            Logger.Log(FunctionId.Run_Environment,
-                KeyValueLogMessage.Create(m => m["Version"] = FileVersionInfo.GetVersionInfo(typeof(VisualStudioWorkspace).Assembly.Location).FileVersion));
+            _componentModel = (IComponentModel)await GetServiceAsync(typeof(SComponentModel)).ConfigureAwait(true);
+            Assumes.Present(_componentModel);
 
-            InitializeColors();
+            cancellationToken.ThrowIfCancellationRequested();
+            _workspace = _componentModel.GetService<VisualStudioWorkspace>();
+            var services = _workspace.Services;
 
-            // load some services that have to be loaded in UI thread
-            LoadComponentsInUIContextOnceSolutionFullyLoadedAsync(cancellationToken).Forget();
+            // Initialize ColorScheme support
+            _colorSchemeApplier = _componentModel.GetService<ColorSchemeApplier>();
+            _colorSchemeApplier.Initialize();
 
-            _solutionEventMonitor = new SolutionEventMonitor(_workspace);
+            _solutionEventMonitor = new SolutionEventMonitor(services);
 
-            TrackBulkFileOperations();
+            TrackBulkFileOperations(services);
 
-            var settingsEditorFactory = _componentModel.GetService<SettingsEditorFactory>();
-            RegisterEditorFactory(settingsEditorFactory);
+            RegisterEditorFactory(_componentModel.GetService<SettingsEditorFactory>());
+
+            InitializeOnBackgroundThreadAsync(services, _componentModel, telemetrySession, cancellationToken).Forget();
+
+            // load some services that have to be loaded in UI thread
+            LoadComponentsInUIContextOnceSolutionFullyLoadedAsync(cancellationToken).Forget();
         }
 
-        private async Task LoadOptionPersistersAsync(IComponentModel componentModel, CancellationToken cancellationToken)
+        private static async Task InitializeOnBackgroundThreadAsync(HostWorkspaceServices services, IComponentModel componentModel, TelemetrySession telemetrySession, CancellationToken cancellationToken)
         {
             var listenerProvider = componentModel.GetService<IAsynchronousOperationListenerProvider>();
-            using var token = listenerProvider.GetListener(FeatureAttribute.Workspace).BeginAsyncOperation(nameof(LoadOptionPersistersAsync));
+            using var token = listenerProvider.GetListener(FeatureAttribute.Workspace).BeginAsyncOperation(nameof(InitializeOnBackgroundThreadAsync));
 
             // Switch to a background thread to ensure assembly loads don't show up as UI delays attributed to
             // InitializeAsync.
             await TaskScheduler.Default;
 
-            var persisterProviders = componentModel.GetExtensions<IOptionPersisterProvider>().ToImmutableArray();
+            var telemetryService = (VisualStudioWorkspaceTelemetryService)services.GetRequiredService<IWorkspaceTelemetryService>();
+            telemetryService.InitializeTelemetrySession(telemetrySession);
```

Comment:
Does this mean there's a window of time where our package can be loaded but we haven't initialized telemetry yet? Is that problematic? I would have imagined we just would have done a switch back to the background thread in the Initialize() and continued to initialize there.

Reply:
I don't see `Initialize()`. AsyncPackage doesn't have it.

Reply:
> Does this mean there's a window of time where our package can be loaded but we haven't initialized telemetry yet? Is that problematic

Yes. I don't think it's problematic. The same happens for OOP.

Reply:
Sorry I meant InitializeAsync.

Reply:
Yes, we could inline this method, but the same pattern was already used and perhaps it's better to clearly separate UI thread init from background thread.

Reply:
> Does this mean there's a window of time where our package can be loaded but we haven't initialized telemetry yet?

There is a _much_ bigger window of time where our package is not loaded at all (many IDE operations do not require the package to be loaded).

Reply:
That issue was already present. We initialize the telemetry service in RoslynPackage.InitializeAsync, currently on UI thread.

Perhaps we can move the initialization to VIsualStudioWorkspaceImpl.InitializeUIAffinitizedServicesAsync?

Reply:
Right, my point is this PR doesn't introduce (or significantly exacerbate) the underlying problem.